src/e2e/java/teammates/e2e/cases/e2e/InstructorCoursesPageE2ETest.java

```diff
@@ -0,0 +1,192 @@
+        coursesPage.verifyActiveCoursesDetails(activeCoursesWithNewCourse);
+        assertFalse(isCourseInRecycleBin(newCourse.getId()));
+
+        ______TS("move archived course to recycle bin");
+        coursesPage.archiveCourse(newCourse.getId());
+        newCourse.deletedAt = Instant.now();
+        coursesPage.moveArchivedCourseToRecycleBin(newCourse.getId());
+
+        coursesPage.verifyStatusMessage("The course " + newCourse.getId() + " has been deleted. "
+                + "You can restore it from the Recycle Bin manually.");
+        coursesPage.verifyNumArchivedCourses(1);
+        coursesPage.verifyDeletedCoursesDetails(deletedCoursesWithNewCourse);
+        assertTrue(isCourseInRecycleBin(newCourse.getId()));
+
+        ______TS("restore archived course");
+        newCourse.deletedAt = null;
+        coursesPage.restoreCourse(newCourse.getId());
+
+        coursesPage.verifyStatusMessage("The course " + newCourse.getId() + " has been restored.");
+        coursesPage.verifyNumDeletedCourses(1);
+        coursesPage.verifyArchivedCoursesDetails(archivedCoursesWithNewCourse);
+        assertFalse(isCourseInRecycleBin(newCourse.getId()));
+        verifyCourseArchivedInDatastore(instructorId, newCourse);
+
+        ______TS("permanently delete course");
+        coursesPage.moveArchivedCourseToRecycleBin(newCourse.getId());
+        coursesPage.deleteCourse(newCourse.getId());
+
+        coursesPage.verifyStatusMessage("The course " + newCourse.getId()
+                + " has been permanently deleted.");
+        coursesPage.verifyNumDeletedCourses(1);
+        verifyAbsentInDatastore(newCourse);
+
+        ______TS("restore all");
+        coursesPage.moveArchivedCourseToRecycleBin(courses[1].getId());
+        CourseAttributes[] activeCoursesWithRestored = { courses[0], courses[2] };
+        coursesPage.restoreAllCourses();
+
+        coursesPage.verifyStatusMessage("All courses have been restored.");
+        coursesPage.sortByCourseId();
+        coursesPage.verifyActiveCoursesDetails(activeCoursesWithRestored);
+        coursesPage.verifyArchivedCoursesDetails(archivedCourses);
+        coursesPage.verifyNumDeletedCourses(0);
+        assertFalse(isCourseInRecycleBin(courses[1].getId()));
+        assertFalse(isCourseInRecycleBin(courses[2].getId()));
+
+        ______TS("permanently delete all");
+        coursesPage.moveArchivedCourseToRecycleBin(courses[1].getId());
+        coursesPage.moveCourseToRecycleBin(courses[2].getId());
+        coursesPage.deleteAllCourses();
+
+        coursesPage.verifyStatusMessage("All courses have been permanently deleted.");
+        coursesPage.verifyNumActiveCourses(1);
+        coursesPage.verifyNumArchivedCourses(0);
+        coursesPage.verifyNumDeletedCourses(0);
+        verifyAbsentInDatastore(courses[1]);
+        verifyAbsentInDatastore(courses[2]);
+
+    }
+
+    private void verifyCourseArchivedInDatastore(String instructorId, CourseAttributes course) {
```

Comment:
Why are these methods here instead of in backdoor? (just like `verifyPresentInDatastore` for example)

Reply:
`verifyPresentInDatastore` is actually in `BaseTestCaseWithDatastoreAccess` which is one of the ancestors of all E2E test classes. However, I believe that the courses page test is the only test that will require the verification of archived courses in the datastore and hence I left it at that level.